ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,192 @@
+            final Map<String, Double> consumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && !m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> restoreConsumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> producerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(PRODUCER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> threadMetrics = allMetrics.stream()
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            metrics.contains(m.metricName().name()))
+                    .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final List<Metric> startTimeList = allMetrics.stream()
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            m.metricName().name().equals("thread-start-time"))
+                    .collect(Collectors.toList());
+            final Long threadStartTime = startTimeList.size() != 0 ? (Long) startTimeList.get(0).metricValue() : 0L;
+
+            double blockedTime = 0.0;
+            if (threadStartTime > windowStart) {
+                blockedTime += threadStartTime - windowStart;
+            }
+            final double newFlushTime = threadMetrics.getOrDefault("flush-time-total", windowSize);
+            double newConsumerIOTime = consumerMetrics.getOrDefault("io-waittime-total", windowSize)
+                + consumerMetrics.getOrDefault("iotime-total", windowSize)
+                + restoreConsumerMetrics.getOrDefault("io-waittime-total", windowSize)
```

Comment:
Okay, I think I miss sth. here since these consumer metrics are not there in AK, did we add them ourselves? If yes could you point me to these metrics?

Reply:
It's a bit hard to tell how the metric names are assembled, but these are the metrics collected here:
https://github.com/confluentinc/kafka/blob/4fea624446b3e3fc889ca8f5938de2497ed5e091/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L467
https://github.com/confluentinc/kafka/blob/4fea624446b3e3fc889ca8f5938de2497ed5e091/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L492

They measure the total time spent in `select` waiting for data, and reading/writing data from/to the network, respectively.

We went with these lower-level metrics because the original metrics we tried to use, which wrapped all of `poll` and `send` were too coarse-grained. They included significant time spent calling interceptors and doing compression (for `send`), which we don't want to count as blocked, because we're using CPU to run the interceptors/compress.

Reply:
I know these metrics, but the names are not exactly the same: e.g. there are `io-wait-time-ns-avg` / `io-time-ns-avg` but there are no `io-waittime-total` / `iotime-total`

Reply:
It gets created by `Meter` (https://github.com/confluentinc/kafka/blob/4fea624446b3e3fc889ca8f5938de2497ed5e091/clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java). The name is computed here: https://github.com/confluentinc/kafka/blob/4fea624446b3e3fc889ca8f5938de2497ed5e091/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L1279